src/backend/vulkan/src/device.rs

```diff
@@ -16,13 +16,459 @@ use hal::{
 use std::borrow::Borrow;
 use std::ffi::CString;
 use std::ops::Range;
+use std::pin::Pin;
 use std::sync::Arc;
 use std::{mem, ptr};
 
 use crate::pool::RawCommandPool;
 use crate::{conv, native as n, window as w};
 use crate::{Backend as B, Device};
 
+struct GraphicsPipelineInfoBuf {
+    // 10 is the max amount of dynamic states
+    dynamic_states: SmallVec<[vk::DynamicState; 10]>,
+
+    // 5 is the amount of stages
+    c_strings: SmallVec<[CString; 5]>,
+    stages: SmallVec<[vk::PipelineShaderStageCreateInfo; 5]>,
+    specializations: SmallVec<[vk::SpecializationInfo; 5]>,
+    specialization_entries: SmallVec<[SmallVec<[vk::SpecializationMapEntry; 4]>; 5]>,
+
+    vertex_bindings: Vec<vk::VertexInputBindingDescription>,
+    vertex_attributes: Vec<vk::VertexInputAttributeDescription>,
+    blend_states: Vec<vk::PipelineColorBlendAttachmentState>,
+    sample_mask: [u32; 2],
+    vertex_input_state: vk::PipelineVertexInputStateCreateInfo,
+    input_assembly_state: vk::PipelineInputAssemblyStateCreateInfo,
+    tessellation_state: Option<vk::PipelineTessellationStateCreateInfo>,
+    viewport_state: vk::PipelineViewportStateCreateInfo,
+    rasterization_state: vk::PipelineRasterizationStateCreateInfo,
+    multisample_state: vk::PipelineMultisampleStateCreateInfo,
+    depth_stencil_state: vk::PipelineDepthStencilStateCreateInfo,
+    color_blend_state: vk::PipelineColorBlendStateCreateInfo,
+    pipeline_dynamic_state: vk::PipelineDynamicStateCreateInfo,
+    viewport: vk::Viewport,
+    scissor: vk::Rect2D,
+}
+impl GraphicsPipelineInfoBuf {
+    fn new<'a>(device: &Device, desc: &pso::GraphicsPipelineDesc<'a, B>) -> Pin<Box<Self>> {
```

Comment:
Does it have to be in a box? This seems like a regression from the original code that didn't box it.